### PR TITLE
Fix handle offset commit fail when offset producer closed

### LINKED ISSUE #1211: [BUG] KoP commit offset failed when doing unload
**Describe the bug**

Currently, when the bundle is to unload, the GroupMetadataManager's producer will be closed, when the KoP receives a commit offset request, then this request will fail with an unknown error. 

Client-side log:
```
7:29:51.440 [pool-3-thread-1] ERROR io.streamnative.tests.pulsar.chaos.test.DurationTestBase - Failed to acknowledge message. offset: 42, receiveIndex: 339, lastReceiveIndex: -1, errMsg: Unexpected error in commit: The server experienced an unexpected error when processing the request.
org.apache.kafka.common.KafkaException: Unexpected error in commit: The server experienced an unexpected error when processing the request.
	at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator$OffsetCommitResponseHandler.handle(ConsumerCoordinator.java:1260) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator$OffsetCommitResponseHandler.handle(ConsumerCoordinator.java:1163) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.internals.AbstractCoordinator$CoordinatorResponseHandler.onSuccess(AbstractCoordinator.java:1177) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.internals.AbstractCoordinator$CoordinatorResponseHandler.onSuccess(AbstractCoordinator.java:1152) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.internals.RequestFuture$1.onSuccess(RequestFuture.java:206) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.internals.RequestFuture.fireSuccess(RequestFuture.java:169) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.internals.RequestFuture.complete(RequestFuture.java:129) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient$RequestFutureCompletionHandler.fireCompletion(ConsumerNetworkClient.java:602) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.firePendingCompletedRequests(ConsumerNetworkClient.java:412) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:297) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:236) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:215) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.commitOffsetsSync(ConsumerCoordinator.java:1005) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.KafkaConsumer.commitSync(KafkaConsumer.java:1490) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.KafkaConsumer.commitSync(KafkaConsumer.java:1390) ~[kafka-clients-2.8.0.jar:?]
	at org.apache.kafka.clients.consumer.KafkaConsumer.commitSync(KafkaConsumer.java:1347) ~[kafka-clients-2.8.0.jar:?]
	at io.streamnative.tests.pulsar.chaos.test.KoPMessagingAtLeastOnceTest.lambda$consumeMsg$0(KoPMessagingAtLeastOnceTest.java:251) ~[test-classes/:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]
```

```
07:29:51.341 [Thread-8850] ERROR io.streamnative.pulsar.handlers.kop.coordinator.group.GroupMetadataManager - Offset commit {persistent://kop/semantic/at-least-once-akhwz-0=[OffsetMetadata[40,NO_METADATA],CommitTime 1648106991252,ExpirationTime 1648366191252], persistent://kop/semantic/at-least-once-akhwz-1=[OffsetMetadata[29,NO_METADATA],CommitTime 1648106991252,ExpirationTime 1648366191252], persistent://kop/semantic/at-least-once-akhwz-2=[OffsetMetadata[41,NO_METADATA],CommitTime 1648106991252,ExpirationTime 1648366191252], persistent://kop/semantic/at-least-once-akhwz-3=[OffsetMetadata[40,NO_METADATA],CommitTime 1648106991252,ExpirationTime 1648366191252], persistent://kop/semantic/at-least-once-akhwz-4=[OffsetMetadata[31,NO_METADATA],CommitTime 1648106991252,ExpirationTime 1648366191252], persistent://kop/semantic/at-least-once-akhwz-5=[OffsetMetadata[38,NO_METADATA],CommitTime 1648106991252,ExpirationTime 1648366191252], persistent://kop/semantic/at-least-once-akhwz-6=[OffsetMetadata[43,NO_METADATA],CommitTime 1648106991252,ExpirationTime 1648366191252], persistent://kop/semantic/at-least-once-akhwz-7=[OffsetMetadata[40,NO_METADATA],CommitTime 1648106991252,ExpirationTime 1648366191252], persistent://kop/semantic/at-least-once-akhwz-8=[OffsetMetadata[42,NO_METADATA],CommitTime 1648106991252,ExpirationTime 1648366191252]} from group group1, consumer DemoKafkaOnPulsarConsumerSSL-2b55510f-d7e3-463d-ad32-cdc6a52bab3c with generation 6 failed when appending to log due to 
java.util.concurrent.CompletionException: org.apache.pulsar.client.api.PulsarClientException$AlreadyClosedException: The producer chaos-pulsar-28-7-34 of the topic public/__kafka/__consumer_offsets-partition-10 was already closed when closing the producers
	at java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:367) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:376) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniRelay.tryFire(CompletableFuture.java:1019) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088) ~[?:?]
	at org.apache.pulsar.client.impl.ProducerImpl$1.sendComplete(ProducerImpl.java:347) ~[io.streamnative-pulsar-client-original-2.9.2.9.jar:2.9.2.9]
	at org.apache.pulsar.client.impl.ProducerImpl$OpSendMsg.sendComplete(ProducerImpl.java:1291) ~[io.streamnative-pulsar-client-original-2.9.2.9.jar:2.9.2.9]
	at org.apache.pulsar.client.impl.ProducerImpl.lambda$failPendingMessages$18(ProducerImpl.java:1827) ~[io.streamnative-pulsar-client-original-2.9.2.9.jar:2.9.2.9]
	at java.util.ArrayDeque.forEach(ArrayDeque.java:889) ~[?:?]
	at org.apache.pulsar.client.impl.ProducerImpl$OpSendMsgQueue.forEach(ProducerImpl.java:1370) ~[io.streamnative-pulsar-client-original-2.9.2.9.jar:2.9.2.9]
	at org.apache.pulsar.client.impl.ProducerImpl.failPendingMessages(ProducerImpl.java:1817) ~[io.streamnative-pulsar-client-original-2.9.2.9.jar:2.9.2.9]
	at org.apache.pulsar.client.impl.ProducerImpl.closeAndClearPendingMessages(ProducerImpl.java:964) ~[io.streamnative-pulsar-client-original-2.9.2.9.jar:2.9.2.9]
	at org.apache.pulsar.client.impl.ProducerImpl.closeAsync(ProducerImpl.java:931) ~[io.streamnative-pulsar-client-original-2.9.2.9.jar:2.9.2.9]
	at io.streamnative.pulsar.handlers.kop.coordinator.group.GroupMetadataManager.lambda$null$74(GroupMetadataManager.java:1169) ~[C44vRLkVGY-jWrCmMdzE9A/:?]
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642) [?:?]
	at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:478) [?:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: org.apache.pulsar.client.api.PulsarClientException$AlreadyClosedException: The producer chaos-pulsar-28-7-34 of the topic public/__kafka/__consumer_offsets-partition-10 was already closed when closing the producers
	at org.apache.pulsar.client.impl.ProducerImpl.closeAndClearPendingMessages(ProducerImpl.java:961) ~[io.streamnative-pulsar-client-original-2.9.2.9.jar:2.9.2.9]
	... 5 more
```

### PR DESCRIPTION
**Pull request by Demogorgon314**
Fixes #1211

### Motivation
Currently, when we get offset producer, then the unload happened, the current producer we get might already close, so the commit offset request will fail.

### Modifications

1. When the producer closed, we should return Errors.NOT_COORDINATOR.
2. Add a units test to cover it.

### Documentation

Check the box below.

Need to update docs? 

- [ ] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [x] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

**Review by eolivelli — APPROVED** [review]
LGTM

good catch !

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandlerTest.java [hunk 1]
```diff
@@ -1175,6 +1180,65 @@ public void testBrokerHandleTopicMetadataRequestAllowAutoTopicCreation(boolean b
...
+
+        // Make sure only close once.
+        final AtomicBoolean flag = new AtomicBoolean(true);
+
+        while (fetchMessages < numMessages) {
+            try {
+                ConsumerRecords<String, String> records = consumer.poll(Duration.ofMillis(1000));
+                records.forEach(record -> {
+                    record.offset();
+                    record.partition();
```
Comment (BewareMyPower) [inline tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandlerTest.java:1220]: Remove unused lines.

```suggestion
```

**Review by BewareMyPower — CHANGES_REQUESTED** [review]

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandlerTest.java [hunk 2]
```diff
@@ -1175,6 +1180,65 @@ public void testBrokerHandleTopicMetadataRequestAllowAutoTopicCreation(boolean b
...
+                            } catch (PulsarClientException | InterruptedException | ExecutionException e) {
+                                log.error("Close offset producer failed.");
+                            }
+                        });
+                        flag.set(false);
+                    }
+                });
+                fetchMessages += records.count();
+            } catch (KafkaException ex) {
+                fail("Should not have kafka exception.");
```
Comment (BewareMyPower) [inline tests/src/test/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandlerTest.java:1236]: Maybe it's better to print the `ex` before `fail`.

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/GroupMetadataManager.java [hunk 3]
```diff
@@ -582,6 +583,9 @@ public GroupMetadata addGroup(GroupMetadata group) {
                                 + " when appending to log due to ",
                         filteredOffsetMetadata, group.groupId(), consumerId, group.generationId(), cause);
 
+                if (cause.getCause() instanceof PulsarClientException.AlreadyClosedException) {
```
Comment (BewareMyPower) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/GroupMetadataManager.java:586]: I think we can try to remove the producer from the cache.

```java
                    offsetsProducers.remove(partitionFor(group.groupId()))
```

After that, there will be no need to call `handler.getGroupCoordinator().getOffsetsProducers().clear()` in the test.

Comment (eolivelli) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/GroupMetadataManager.java:586]: +1 good catch

Comment (Demogorgon314) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/GroupMetadataManager.java:586]: Good catch. Addressed.

**Review by eolivelli — APPROVED** [review]

**Review by wenbingshen — APPROVED** [review]
Good work!

**Pull request merged**